### ISSUE TITLE
typo and missing return statements

### ISSUE THREAD

**Issue by BrettRyland** [pytorch/pytorch#21777]
## 🐛 Bug

1. On the current master (556af7c19), building causes a lot of unnecessary spam from `ThrowInTheTowelIfGradientIsCalled` and `GradientNotImplementedYet`.
2. A lambda in the ideep third party module is missing a trailing `;`. This appears to be corrected in the upstream version of ideep.
3. A return statement is missing in `Eigen/src/Core/arch/GPU/PacketMathHalf.h` in the eigen third party module.

The attached patches clean these up.

[pytorch_main.txt](https://github.com/pytorch/pytorch/files/3290023/pytorch_main.txt)
[pytorch_third_party_eigen.txt](https://github.com/pytorch/pytorch/files/3290024/pytorch_third_party_eigen.txt)
[pytorch_third_party_ideep.txt](https://github.com/pytorch/pytorch/files/3290025/pytorch_third_party_ideep.txt)

Notes: Building from source on Debian (mostly unstable, some experimental) with a fresh clone; building directly in `pytorch/build` with `ccmake ..` (to specify CUDA_HOST_COMPILER and some other variables) and `make -j8` prior to installing with `python3 setup.py install`.

**Comment by zhangguanheng66** [comment]
@BrettRyland Could you submit a PR?

**Comment by colesbury** [comment]
I believe the ThrowInTheTowelIfGradientIsCalled warnings are because of CAFFE_ENFORCE(false) now using `C10_UNLIKELY`

cc @smessmer

**Comment by BrettRyland** [comment]
Yeah, I doubt the patches I made are anywhere near the best solutions (despite compiling fine). Looking at upstream for ideep and eigen, these fixes are made in commits [d72721b](https://github.com/intel/ideep/commit/d72721ba44624dd9b06678f1cb6400a6a8c5e647#diff-89f4e64b25c453aac8d8057b9ddb8d78) and [5713d8b28](https://github.com/eigenteam/eigen-git-mirror/commit/5713d8b28ff946276b493736498cb231e73f46ac#diff-a0a131f867cd508ea3aa4ad4495179a7), so really, those third party submodules should just be updated from upstream.

**Comment by colesbury** [comment]
@smesser can you fix this?

**Comment by SMesser** [comment]
Um. Maybe? I haven't done hardware GPU acceleration before and am not familiar with the contributors, project, or problem... Are you sure you tagged the person you intended?

**Comment by SMesser** [comment]
Also, I'm getting notifications for @smessmer, even though I'm @smesser...

**Comment by colesbury** [comment]
Sorry, typo... (I edited the tag)

**Comment by smessmer** [comment]
Don't think it's because of `C10_UNLIKELY`, more likely because `CAFFE_ENFORCE(false, ...)` doesn't work with a `[[noreturn]]` attribute. Afaik, we said we're going to introduce something like `TORCH_CHECK_FAIL()` and `TORCH_INTERNAL_ASSERT_FAIL()`, cc @ezyang @dzhulgakov

**Comment by smessmer** [comment]
the "missing return statement" warning should be fixed in https://github.com/pytorch/pytorch/pull/22216
